File: backend/routes/bookings.py
"""
TravelVista — Bookings Routes
"""
import time
from datetime import date
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import get_db, booking_to_dict

bookings_bp = Blueprint('bookings', __name__, url_prefix='/api/bookings')


def _is_admin(uid):
    conn = get_db()
    row = conn.execute("SELECT role FROM users WHERE id = ?", (uid,)).fetchone()
    conn.close()
    return row and row['role'] == 'admin'


@bookings_bp.route('', methods=['GET'])
@jwt_required()
def list_bookings():
    """Admin: list all bookings."""
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    rows = conn.execute("SELECT * FROM bookings ORDER BY created_at DESC").fetchall()
    conn.close()
    return jsonify([booking_to_dict(r) for r in rows])


@bookings_bp.route('/my', methods=['GET'])
@jwt_required()
def my_bookings():
    """User: list own bookings."""
    uid = get_jwt_identity()
    conn = get_db()
    rows = conn.execute(
        "SELECT * FROM bookings WHERE user_id = ? ORDER BY created_at DESC", (uid,)
    ).fetchall()
    conn.close()
    return jsonify([booking_to_dict(r) for r in rows])


@bookings_bp.route('', methods=['POST'])
@jwt_required()
def create_booking():
    uid = get_jwt_identity()
    data = request.get_json() or {}

    booking_id = 'bk-' + str(int(time.time() * 1000))
    conn = get_db()
    conn.execute(
        """INSERT INTO bookings
           (id,user_id,package_id,package_title,travel_date,persons,total_amount,
            status,payment_status,payment_method,created_at)
           VALUES (?,?,?,?,?,?,?,?,?,?,?)""",
        (
            booking_id,
            uid,
            data.get('packageId', ''),
            data.get('packageTitle', ''),
            data.get('travelDate', ''),
            data.get('persons', 1),
            data.get('totalAmount', 0),
            'pending',
            'pending',
            '',
            date.today().isoformat(),
        )
    )
    conn.commit()
    row = conn.execute("SELECT * FROM bookings WHERE id = ?", (booking_id,)).fetchone()
    conn.close()
    return jsonify(booking_to_dict(row)), 201


@bookings_bp.route('/<bid>/status', methods=['PUT'])
@jwt_required()
def update_status(bid):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    status = data.get('status', '')
    if not status:
        return jsonify({'message': 'Status required'}), 400

    conn = get_db()
    conn.execute("UPDATE bookings SET status = ? WHERE id = ?", (status, bid))
    conn.commit()
    row = conn.execute("SELECT * FROM bookings WHERE id = ?", (bid,)).fetchone()
    conn.close()
    if not row:
        return jsonify({'message': 'Booking not found'}), 404
    return jsonify(booking_to_dict(row))


@bookings_bp.route('/<bid>/pay', methods=['PUT'])
@jwt_required()
def confirm_payment(bid):
    data = request.get_json() or {}
    method = data.get('paymentMethod', 'card')

    conn = get_db()
    conn.execute(
        "UPDATE bookings SET status='confirmed', payment_status='paid', payment_method=? WHERE id=?",
        (method, bid)
    )
    conn.commit()
    row = conn.execute("SELECT * FROM bookings WHERE id = ?", (bid,)).fetchone()
    conn.close()
    if not row:
        return jsonify({'message': 'Booking not found'}), 404
    return jsonify(booking_to_dict(row))


@bookings_bp.route('/<bid>/cancel', methods=['PUT'])
@jwt_required()
def cancel_booking(bid):
    conn = get_db()
    row = conn.execute("SELECT * FROM bookings WHERE id = ?", (bid,)).fetchone()
    if not row:
        conn.close()
        return jsonify({'message': 'Booking not found'}), 404

    new_payment = 'refunded' if row['payment_status'] == 'paid' else 'cancelled'
    conn.execute(
        "UPDATE bookings SET status='cancelled', payment_status=? WHERE id=?",
        (new_payment, bid)
    )
    conn.commit()
    row = conn.execute("SELECT * FROM bookings WHERE id = ?", (bid,)).fetchone()
    conn.close()
    return jsonify(booking_to_dict(row))


@bookings_bp.route('/revenue', methods=['GET'])
@jwt_required()
def revenue():
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    rows = conn.execute(
        "SELECT * FROM bookings WHERE payment_status = 'paid'"
    ).fetchall()
    conn.close()

    total = sum(r['total_amount'] for r in rows)
    months = {}
    for r in rows:
        m = r['created_at'][:7]
        months[m] = months.get(m, 0) + r['total_amount']

    monthly = [{'month': k, 'revenue': v} for k, v in sorted(months.items())]
    return jsonify({'totalRevenue': total, 'monthlyRevenue': monthly})

File: api/bookings.php
<?php
/**
 * TravelVista — Bookings API
 * GET                          — List all bookings (admin)
 * GET    ?action=my            — List current user's bookings
 * POST                         — Create booking
 * PUT    ?id=X&action=status   — Update booking status (admin)
 * PUT    ?id=X&action=pay      — Confirm payment
 * PUT    ?id=X&action=cancel   — Cancel booking
 * GET    ?action=revenue       — Revenue data (admin)
 */

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') { http_response_code(200); exit; }

require_once __DIR__ . '/../config/jwt_helper.php';

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';
$id = $_GET['id'] ?? '';
if (!$action && !$id) {
    $uri = $_SERVER['REQUEST_URI'] ?? '';
    if (preg_match('#/api/bookings/(\d+)/status#', $uri, $m)) { $id = $m[1]; $action = 'status'; }
    elseif (preg_match('#/api/bookings/(\d+)/pay#', $uri, $m)) { $id = $m[1]; $action = 'pay'; }
    elseif (preg_match('#/api/bookings/(\d+)/cancel#', $uri, $m)) { $id = $m[1]; $action = 'cancel'; }
    elseif (preg_match('#/api/bookings/my#', $uri)) { $action = 'my'; }
    elseif (preg_match('#/api/bookings/revenue#', $uri)) { $action = 'revenue'; }
    elseif (preg_match('#/api/bookings/(\d+)#', $uri, $m)) { $id = $m[1]; }
}

// ── GET: All bookings (admin) ──
if ($method === 'GET' && !$action && !$id) {
    require_admin();
    $db = getDB();
    $rows = $db->query("SELECT b.*, u.name as user_name FROM bookings b LEFT JOIN users u ON b.user_id = u.id ORDER BY b.created_at DESC")->fetchAll();
    json_response(array_map('booking_to_dict', $rows));
}

// ── GET: My bookings ──
if ($method === 'GET' && $action === 'my') {
    $userId = require_auth();
    $db = getDB();
    $stmt = $db->prepare("SELECT * FROM bookings WHERE user_id = ? ORDER BY created_at DESC");
    $stmt->execute([$userId]);
    json_response(array_map('booking_to_dict', $stmt->fetchAll()));
}

// ── GET: Revenue (admin) ──
if ($method === 'GET' && $action === 'revenue') {
    require_admin();
    $db = getDB();
    $rows = $db->query("SELECT * FROM bookings WHERE payment_status = 'paid'")->fetchAll();

    $total = 0;
    $months = [];
    foreach ($rows as $r) {
        $total += (float)$r['total_amount'];
        $m = substr($r['created_at'], 0, 7);
        $months[$m] = ($months[$m] ?? 0) + (float)$r['total_amount'];
    }
    ksort($months);
    $monthly = [];
    foreach ($months as $k => $v) {
        $monthly[] = ['month' => $k, 'revenue' => $v];
    }
    json_response(['totalRevenue' => $total, 'monthlyRevenue' => $monthly]);
}

// ── POST: Create booking ──
if ($method === 'POST') {
    $userId = require_auth();
    $data = get_json_body();
    $db = getDB();

    $stmt = $db->prepare(
        "INSERT INTO bookings (user_id, package_id, package_title, travel_date, persons, total_amount, status, payment_status, payment_method, created_at)
         VALUES (?, ?, ?, ?, ?, ?, 'pending', 'pending', '', ?)"
    );
    $stmt->execute([
        $userId,
        $data['packageId'] ?? 0,
        $data['packageTitle'] ?? '',
        $data['travelDate'] ?? date('Y-m-d'),
        $data['persons'] ?? 1,
        $data['totalAmount'] ?? 0,
        date('Y-m-d'),
    ]);

    $bookingId = $db->lastInsertId();
    $stmt = $db->prepare("SELECT * FROM bookings WHERE id = ?");
    $stmt->execute([$bookingId]);
    json_response(booking_to_dict($stmt->fetch()), 201);
}

// ── PUT: Update status (admin) ──
if ($method === 'PUT' && $id && $action === 'status') {
    require_admin();
    $data = get_json_body();
    $status = $data['status'] ?? '';
    if (!$status) json_response(['message' => 'Status required'], 400);

    $db = getDB();
    $db->prepare("UPDATE bookings SET status = ? WHERE id = ?")->execute([$status, $id]);

    $stmt = $db->prepare("SELECT * FROM bookings WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Booking not found'], 404);
    json_response(booking_to_dict($row));
}

// ── PUT: Confirm payment ──
if ($method === 'PUT' && $id && $action === 'pay') {
    require_auth();
    $data = get_json_body();
    $payMethod = $data['paymentMethod'] ?? 'card';

    $db = getDB();
    $db->prepare("UPDATE bookings SET status='confirmed', payment_status='paid', payment_method=? WHERE id=?")
       ->execute([$payMethod, $id]);

    $stmt = $db->prepare("SELECT * FROM bookings WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Booking not found'], 404);
    json_response(booking_to_dict($row));
}

// ── PUT: Cancel booking ──
if ($method === 'PUT' && $id && $action === 'cancel') {
    require_auth();
    $db = getDB();
    $stmt = $db->prepare("SELECT * FROM bookings WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Booking not found'], 404);

    $newPayment = ($row['payment_status'] === 'paid') ? 'refunded' : 'cancelled';
    $db->prepare("UPDATE bookings SET status='cancelled', payment_status=? WHERE id=?")
       ->execute([$newPayment, $id]);

    $stmt = $db->prepare("SELECT * FROM bookings WHERE id = ?");
    $stmt->execute([$id]);
    json_response(booking_to_dict($stmt->fetch()));
}

json_response(['message' => 'Invalid request'], 400);
?>
